### ISSUE TITLE
Pasted text includes CRLF pairs where unneeded

### ISSUE THREAD

**Issue by d-bingham** [microsoft/terminal#1091]
Multiline text pasted from the clipboard includes CRLF pairs in all cases; this is inappropriate for "Unix-space" sessions, such as WSL.

# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.145]
Windows Terminal version (if applicable): 71e19cd82528d66a0a7867cbed85990cfc1685f1
```

# Steps to reproduce

Select multiline text in Terminal.
Copy it (via right-click, ostensibly)
Paste it (again via right-click)

# Expected behavior

When pasting into a Unix-space session -- such as WSL -- pasted text should have a reasonable set of line-ending characters.

# Actual behavior

Line endings are "doubled" on text paste to Unix-space sessions.

**Comment by d-bingham** [comment]
![image](https://user-images.githubusercontent.com/26487560/58742970-17ba0480-83ef-11e9-908c-b29e3fc1384b.png)

**Comment by DHowett-MSFT** [comment]
This is now the master issue for line feeds on copy/paste. We need a holistic solution that addresses this ("...includes too many line breaks") and #1073 ("... doesn't include line breaks"). :grin:

**Comment by conioh** [comment]
@DHowett-MSFT @d-bingham The description talks about WSL, but my issue (#1073) happened with no WSL involvement. It's indeed tricky to reproduce (which I noticed after seeing your comments) but happened without WSL.

**Comment by DHowett-MSFT** [comment]
Indeed. This is more likely related to the console's output mode and what was written to the screen and how it was written instead of any _particular_ tool.

**Comment by conioh** [comment]
So PR #1094 won't fix it. Either said PR doesn't close this issue or #1073 isn't a duplicate of this issue.

**Comment by DHowett-MSFT** [comment]
PR #1094 doesn't fix _both_ issues, but we believe that:

> We need a holistic solution that addresses this ("...includes too many line breaks") and #1073 ("... doesn't include line breaks").

And this is valuable feedback for 1094 :smile:

**Comment by conioh** [comment]
Maybe I'm missing something - I admit that I don't know how all this GitHub thing works - but it says above that:
![image](https://user-images.githubusercontent.com/10606081/58841121-21eb3580-8671-11e9-99c1-6fb835233554.png)

It makes it sound as if once #1094 is approved and merged this issue will be close automatically, and if I understand correctly you agree that it shouldn't.

**Comment by d-bingham** [comment]
I'm not sure there's really a "holistic" solution here, as @conioh's issue in #1073 is that the text _copied_ to the clipboard doesn't include line breaks, and #1094 fixes a problem with _pasted_ text having CRLFs when LFs are appropriate.

**Comment by DHowett-MSFT** [comment]
Oh, I see: I got my copies and pastes mixed up.
Thanks!

**Comment by ivan-section-io** [comment]
For Unix/WSL I believe this should be CRLFs to CRs, and not CRLFs to LFs
re: https://github.com/microsoft/terminal/pull/1094#issuecomment-498967181

**Comment by sandric** [comment]
Sorry to bother, but is there at least any workaround (makes it pretty unusable)? I mean in emacs I can redefine pasting function, is there some piping commands reading clipboard I can run from bash to read clipboard with no doubled newlines?

**Comment by sandric** [comment]
Ok, I'm not sure if thats the place for it to live, but if anyone struggles with it, here's mine workaround using autohotkey program:

```
<^t::
  Var := Clipboard
  Clipboard := RegExReplace(Var, "\r\n?|\n\r?", "`n")
  MouseClick, right
return
```

Now, by pressing LCtrl + t I get normal clipboard output with no doubled newlines, if anyone struggles - workaround for now.

**Comment by alexjmoore** [comment]
Just wondering if there was any update? I see the latest 0.3 preview didn't address this yet.

**Comment by FlipperPA** [comment]
Plus one on this bug fix. In Sublime Text 3 on Windows, I can copy this follow text (which I'm pasting correctly here):

```python
import pyautogui
from time import sleep

for x in range(1, 1000):
    print(x)
```

I can also paste it to Notepad, PuTTy, or any number of other programs successfully.

When I right-click to paste it into WSL-2 Ubuntu 18.04 `emacs`, it doubles carriage returns:

![image](https://user-images.githubusercontent.com/68164/63396073-e0c31180-c393-11e9-9094-00be4be80302.png)

Same with `nano`. And when I paste it to WSL-2 `cmd`:

![image](https://user-images.githubusercontent.com/68164/63396123-05b78480-c394-11e9-8e20-3f42434ad743.png)

Thanks for all your efforts on bringing these features to Windows. It is looking incredibly promising as the development environment of choice moving forward, but we'll have to polish some chrome with issues like these.

**Comment by FlipperPA** [comment]
Just to follow up, this is occurring both in WSL-1 and WSL-2, as another data point.

**Comment by JCMais** [comment]
With nano this seems worse, because looks like it starts to replace the current content on the file, right below the pasting point.

**Comment by genio** [comment]
Please fix this. I cannot explain how frustrating this bug is.

**Comment by dquist** [comment]
Appreciate all the hard work on the terminal, I'm using it for my daily work but this one is particularly frustrating. Is there any workaround?

**Comment by ilirb** [comment]
I agree this is currently the most frustrating issue with Windows Terminal - WSL

> Ok, I'm not sure if thats the place for it to live, but if anyone struggles with it, here's mine workaround using autohotkey program:
> 
> ```
> <^t::
>   Var := Clipboard
>   Clipboard := RegExReplace(Var, "\r\n?|\n\r?", "`n")
>   MouseClick, right
> return
> ```
> 
> Now, by pressing LCtrl + t I get normal clipboard output with no doubled newlines, if anyone struggles - workaround for now.

Thanks for this, adjusted a bit to be able to use `Ctrl + Shift + V` while in WindowsTerminal, to not interfere with other apps
```
#if WinActive("ahk_exe WindowsTerminal.exe")
    ^+v::
        Var := Clipboard
        Clipboard := RegExReplace(Var, "\r\n?|\n\r?", "`n")
        MouseClick, right
        return
#if
```

**Comment by CoskunSunali** [comment]
There is also another problem where we lose new lines. I don't know if the two might be related.

#52

**Comment by gitfool** [comment]
ConEmu gets this right. I'm reluctant to switch to Microsoft Terminal until this is working; functional copy paste from Windows to Linux and vice versa is crucial!

**Comment by FingerlessGlov3s** [comment]
I'm getting this issue, now its happening on bash.exe as well as the new terminal on 1909.

How's this BUG not been fixed yet?

I was using bash.exe on 1809, and it worked fine. I thought I'd update to 1909, but now my copy n paste is broken for WSL, I'll have to use putty again now.

**Comment by dquist** [comment]
> How's this BUG not been fixed yet?

Windows Terminal is still in public preview.

**Comment by FingerlessGlov3s** [comment]
That it is 😉

Isn't the idea of the preview to get feedback and bugs for the app?

**Comment by tejasvi** [comment]
As of now, is there any workaround while pasting in vim?

**Comment by FingerlessGlov3s** [comment]
> As of now, is there any workaround while pasting in vim?

You can use Xmobaterm to use wsl and that doesn't have the issue.

**Comment by JCMais** [comment]
Thank you everyone!

Glad to see this finally fixed, it was one of the major pain points while using the new Terminal

**Comment by erwa** [comment]
Is this fix available in any release preview version of Terminal? It doesn't appear to be in v0.6.2951.0.

**Comment by zadjii-msft** [comment]
@erwa not yet. v0.6.2951.0 was the "1910" Terminal milestone, and this is fixed for the "1911" milestone. It will be in the next released preview version of the Terminal, sometime later in November.

**Comment by erwa** [comment]
Gotcha, thanks for the ETA, Mike. I eagerly await the next preview release :).

**Comment by erwa** [comment]
YESS!! Finally!! I can confirm this issue is fixed in the new preview release. Thank you for fixing this!

**Comment by erwa** [comment]
One thing I prefer about using the Windows Terminal over the Ubuntu Shell is Windows Terminal has tabs support built-in whereas the Ubuntu Shell does not (I end up using tmux instead).

**Comment by paul-michalik** [comment]
To be honest I am running into these issues since update to 0.10.781.0.

**Comment by DHowett-MSFT** [comment]
Commenting on a closed issue is not a very good way to get help. Would you mind putting together a more robust bug report so that we can try to reproduce it?

**Comment by paul-michalik** [comment]
@DHowett-MSFT Yep, will do.

**Comment by ThomasLachaux** [comment]
I have the same problem @paul-michalik encounters
When I copy a content that can't fit my window such as an ssh public key on another terminal or even an application like notepad, newlines are inserted where it shouldn't.

I am also an the version 0.10.781.0

**Comment by DHowett-MSFT** [comment]
You're hitting #5113, which is fixed pending release.
